src/core/components/dag.js

```diff
@@ -73,6 +75,30 @@ module.exports = function dag (self) {
         self._ipldResolver.treeStream(cid, path, options),
         pull.collect(callback)
       )
+    }),
+
+    getRecursive: promisify((multihash, callback) => {
```

Comment:
This getRecursive is only working for `ipld-dag-pb` right now. Either make it an internal method by prefixing with a `_` and add a comment saying just that or move it to https://github.com/ipld/js-ipld-resolver and leverage the `.tree` and `.isLink` API calls to make it agnostic to any kind of IPLD Format

Reply:
I'll see if I can move it

Reply:
@diasdavid Should the existing treeStream method already do what I'm trying to do here, if the recursive: true option is passed? I tried testing it with a CID for [this directory](https://ipfs.io/ipfs/QmUhUuiTKkkK8J6JZ9zmj8iNHPuNfGYcszgRumzhHBxEEU) but I'm getting an error in the recursive part. Specifically this [`new CID(...)`](https://github.com/ipld/js-ipld-resolver/blob/master/src/index.js#L314) errors because `paths` includes separate objects for link name, size and hash that all have `p.link` (so it tries to recurse all of them, when only the hash ones should be recursed). Is it a bug that the hashes array is treating these as separate links? Or should the recursion check specifically for the "hash" links?

```
[
   {
      "path":"Links/0/Name",
      "link":{
         "/":"about"
      }
   },
   {
      "path":"Links/0/Tsize",
      "link":{
         "/":1688
      }
   },
   {
      "path":"Links/0/Hash",
      "link":{
         "/":"QmZTR5bcpQD7cFgTorqxZDYaew1Wqgfbd2ud9QqGPAkK2V"
      }
   }
]
```

Reply:
that definitely looks like a bug as it _brings more info than what it promises_. The method was created just to return all the paths. That `link` property is just a left over.

Reply:
@diasdavid Based on [these](https://github.com/ipld/interface-ipld-format/pull/5) [discussions](https://github.com/ipld/js-ipld-dag-pb/pull/29) it looks like the presence of separate "paths" for name, size and hash for each link is correct/necessary, and the bug is that the isLink method (formerly isCID) was meant to identify specifically the "hash" paths but is actually more permissive. The tests pass because the [non-valid CID case](https://github.com/ipld/js-ipld-dag-pb/blob/master/test/resolver.spec.js#L230) is checking an [empty value](https://github.com/ipld/js-ipld-dag-pb/blob/master/test/resolver.spec.js#L25), which happens to return `false` correctly, whereas other non-empty non-CID values would not. I'll make another PR to address that and update the test.